url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw5/SpellCheck.java

```diff
@@ -0,0 +1,273 @@
+						myprob += both;
+					}
+					else
+					{
+						both = predict * update;					
+						myprob *= both;
+					}
+					
+					//System.out.println(both);
+					//System.out.println(myprob + " * "+predict+" * " + update + " = " + myprob * predict * update);
+
+					nn.updateMax(tostate, fromstate, myprob);
+					if (t==0) nn.parent=null; //if we're at zero make sure no parent
+				}
+				//transitionModel.setLast(e);
+				if (nn.state!=null) //TODO: this is bad, this means that we aren't figuring out a path at every step
+					nppaths.put(tostate, nn);
+										
+			}
+			if (t==0) normalize(nppaths); // normalize the first step ignore the rest
+			System.out.println(nppaths);
+			ppaths = nppaths;
+			t++;
+		}
+		
+		
+		//choose
+		if (t==observations.length)
+		{
+			System.out.println("most probable sequence:");
+			Node max;
+			if (logOn)
+				max = new Node(null, null, Math.log(-0), true);
+			else
+				max = new Node(null, null, 0d);
+			for (Character cstate : states)
+			{
+				String state = cstate.toString();
+				Node from = ppaths.get(state);
+				max.updateMax(from.state, from.parent, from.probability);
+			}
+			System.out.println(max);
+		}
+		
+	}
+	
+	
+	
+	/**
+	 * normalizez
+	 * @param in
+	 * @return
+	 */
+	private void normalize(HashMap<String, Node> in)
+	{
+		if (logOn)
+		{
+			Double tots = 0.0;
+			for (Node d : in.values())
+			{
+				tots = tots + Math.exp(d.probability);
+			}
+			
+			Double alpha = 1.0/tots;
+			
+			for (Node s : in.values())
+			{
+				s.probability = Math.log(Math.exp(s.probability) * alpha);
+			}
+		}
+		else
+		{
+			Double tots = 0.0;
+			for (Node d : in.values())
+			{
+				tots = tots + d.probability;
+			}
+			
+			Double alpha = 1.0/tots;
+			
+			for (Node s : in.values())
+			{
+				s.probability = s.probability * alpha;
+			}
+		}
+	}
+
+	
+	
+	
+	/**
+	 * this is the initial probabiltiy
+	 * simply 1/27th for any letter for now
+	 */
+	private static class initial_probability
+	{
+		
+		static Double get(String state)
+		{
+			return 1d/27d;
+		}
+	}
+	
+	/**
+	 * this is the transition model
+	 * stores the last few characters and give P(A | BC) probability
+	 * TODO: this doesn't work right
+	 */
+	private class TransitionModel
+	{
+		LinkedList<String> last;
+		
+		TransitionModel()
+		{
+			last = new LinkedList<String>();
+		}
+		
+		void setLast(String l)
+		{
+			last.addLast(l);
+		}
+		
+		Double get(String fromstate, String tostate)
+		{
+			Double t = tchars.bigramCondProb(tostate.charAt(0), last.getLast() + fromstate);
+			//System.out.println("P("+ tostate.charAt(0)+"|"+last.getLast() + fromstate + ") = " + t);
+			//System.out.println(last);
+			if (logOn)
+				t = (t==0d) ? Math.log(-0) : t;
+			return t;
+		}
+
+	}
+	
+	
+	
+	/**
+	 * this is the sensor model, models the error probable in the text
+	 *
+	 */
+	private class SensorModel
+	{
+		
+		SensorModel()
+		{
+		}
+		
+		Double get(String tostate, String e)
+		{
+			if (e == tostate)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.